### PR TITLE
Consolidate scripts

### PR DESCRIPTION
**Pull request by Helcaraxan**
Introducing the appropriate CI helper scripts to ensure that when regenerating the API packages we are always using an appropriate version of Bazel, regardless of what you have installed locally. These helper scripts are identical to what we use elsewhere in other code-bases.

Also doing a general pass on the other scripts to align their Bash style with best-practices.

### CODE REVIEW

**Comment by Helcaraxan** [comment]
Friendly ping on this for review.

File: scripts/nightly.sh [hunk 1]
```diff
@@ -1,24 +1,37 @@
 #!/usr/bin/env bash
 
-set -euo pipefail
+set -e -u -o pipefail
 
-REPO_ROOT=$(cd "$(dirname "${BASH_SOURCE[0]}")/.." && pwd)
-SDK_VERSION_NIGHTLY=999.9.9-NIGHTLY
-DOCKER_IMAGE=platform-sdk/csharp:$SDK_VERSION_NIGHTLY
+REPO_ROOT="$(dirname "${BASH_SOURCE[0]}")/.."
```
Comment (RNabel) [inline scripts/nightly.sh]: In most of the `everything` codebase we use:
```
cd "$(dirname "$0")/../../../"
```
so maybe change this to
```
REPO_ROOT="$(dirname "$0")/.."
```
for consistency.

File: scripts/nightly.sh [hunk 2]
```diff
@@ -1,24 +1,37 @@
...
+echo "--- Regenerating APIs"
+echo "Skipping until bazel is available on the agent"
+# scripts/generateapis.sh
 
 echo "--- Preparing docker image for nightly"
-docker build --build-arg IMPROBABLE_REFRESH_TOKEN=$IMPROBABLE_REFRESH_TOKEN --build-arg SDK_VERSION=$SDK_VERSION_NIGHTLY -t $DOCKER_IMAGE $REPO_ROOT
+docker build \
+  --build-arg "IMPROBABLE_REFRESH_TOKEN=${IMPROBABLE_REFRESH_TOKEN}" \
+  --build-arg "SDK_VERSION=${SDK_VERSION_NIGHTLY}" \
+  -t "${DOCKER_IMAGE}" \
```
Comment (RNabel) [inline scripts/nightly.sh]: Replace `-t` with `--tty`

Comment (RNabel) [reply scripts/nightly.sh]: Same below.

File: scripts/nightly.sh (hunk 1 above)
Comment (RNabel) [reply scripts/nightly.sh]: Same below

File: scripts/premerge.sh [hunk 3]
```diff
@@ -1,24 +1,37 @@
 #!/usr/bin/env bash
```
Comment (RNabel) [inline scripts/premerge.sh]: What is the difference between this file and `nightly.sh`? They seem to share 90%+ of logic.

File: scripts/runtests.sh [hunk 4]
```diff
@@ -1,25 +1,26 @@
 #!/usr/bin/env bash
 
-set -e
+set -e -u -o pipefail
 
-if [[ $# -ne 1 ]]; then
+if [[ "$#" -ne 1 ]]; then
```
Comment (RNabel) [inline scripts/runtests.sh]: This can only be a number, so doesn't _need_ to be quoted.

File: scripts/runtests.sh [hunk 5]
```diff
@@ -1,25 +1,26 @@
...
-docker run $DOCKER_IMAGE /bin/bash -c "cd examples/ServiceAccountMaintenance && dotnet run --no-build --configuration Release"
+docker run "${DOCKER_IMAGE}" /bin/bash -c "cd examples/ServiceAccountMaintenance && dotnet run --no-build --configuration Release"
 
 echo "--- Running Replicate State Scenario"
-docker run $DOCKER_IMAGE /bin/bash -c "cd examples && spatial service start --main_config=blank_project/spatialos.json && cd ReplicateState && dotnet run --no-build --configuration Release"
+docker run "${DOCKER_IMAGE}" /bin/bash -c "cd examples && spatial service start --main_config=blank_project/spatialos.json && cd ReplicateState && dotnet run --no-build --configuration Release"
 
 echo "--- Running BYO Auth Scenario"
-docker run $DOCKER_IMAGE /bin/bash -c "cd examples && cd BYOAuthFlow && dotnet run --no-build --configuration Release"
+docker run "${DOCKER_IMAGE}" /bin/bash -c "cd examples && cd BYOAuthFlow && dotnet run --no-build --configuration Release"
```
Comment (RNabel) [inline scripts/runtests.sh]: I'd recommend using `--workdir="/workspace/examples/BYOAuthFlow"` instead of this chaining of `cd`s.
Same elsewhere.

Comment (RNabel) [reply scripts/runtests.sh]: And in any case, do we know why there are multiple `cd`s in the first place?

File: scripts/generateapis.sh [hunk 6]
```diff
@@ -1,22 +1,25 @@
...
+    PACKAGE="apis/${NAME}_${VERSION}"
 
-    bazel build //$PACKAGE:gapic
-    bazel build //$PACKAGE:grpc
-    rm -rf $PACKAGE/*.cs
-    cp -R $REPO_ROOT/bazel-genfiles/$PACKAGE/${NAME}_gapicout/*/*/*.cs $REPO_ROOT/$PACKAGE
-    cp -R $REPO_ROOT/bazel-genfiles/$PACKAGE/*.cs $REPO_ROOT/$PACKAGE
+    runBazel build "//${PACKAGE}:gapic"
+    runBazel build "//${PACKAGE}:grpc"
+    rm -rf "${PACKAGE}/*.cs"
```
[the remark this answers is not in the document]
Comment (Helcaraxan) [reply scripts/generateapis.sh]: Yes indeed. Had missed this occurrence of a quoted glob.

File: scripts/nightly.sh (hunk 1 above)
Comment (Helcaraxan) [reply scripts/nightly.sh]: Applied everywhere.

File: scripts/nightly.sh (hunk 2 above)
Comment (Helcaraxan) [reply scripts/nightly.sh]: Applied everywhere.

File: scripts/premerge.sh (hunk 3 above)
Comment (Helcaraxan) [reply scripts/premerge.sh]: Not entirely sure. @lx223 any details you can give here?

File: scripts/runtests.sh (hunk 4 above)
Comment (Helcaraxan) [reply scripts/runtests.sh]: Yes. Still doing it though as per-style.

File: scripts/runtests.sh (hunk 5 above)
Comment (Helcaraxan) [reply scripts/runtests.sh]: Don't know about the background but will happily use the `--workdir` approach indeed.

**Comment by Helcaraxan** [comment]
@RNabel have indeed addressed the `--workdir` paths issue that we spoke offline about.

**Comment by petemounce** [comment]
@Helcaraxan --tty is I think not in the version of docker in that platform build queue. There are other PRs that update the build queue that might fix the current failure. If those don't work, the latest queue might. If that doesn't work, a new queue could be baked with latest docker; that's in-flight.

**Comment by petemounce** [comment]
Hurrah!

**Pull request merged**